### PR TITLE
gh-103606: Improve error message from logging.config.FileConfig

### PR DESCRIPTION
**Pull request by Agent-Hellboy**
* Issue: gh-103606

### CODE REVIEW

File: Lib/logging/config.py [hunk 1]
```diff
@@ -68,8 +68,9 @@ def fileConfig(fname, defaults=None, disable_existing_loggers=True, encoding=Non
             cp.read_file(fname)
         else:
             encoding = io.text_encoding(encoding)
-            cp.read(fname, encoding=encoding)
-
+            files = cp.read(fname, encoding=encoding)
```
Comment (vsajip) [inline Lib/logging/config.py]: This can be kept in case e.g. there's an invalid file there, but it would IMO be better to check for file existence and bail with a `FileNotFoundError` if there's no such file. That's more informative. Then this error (no parsed filenames) would lead to a message "The configuration file XXX appears not to be in the expected format" or some such.

File: Lib/test/test_logging.py [hunk 2]
```diff
@@ -1756,6 +1756,10 @@ def test_config_set_handler_names(self):
         self.apply_config(test_config)
         self.assertEqual(logging.getLogger().handlers[0].name, 'hand1')
 
+    def test_exception_if_file_not_present(self):
```
Comment (vsajip) [inline Lib/test/test_logging.py]: Perhaps check for (a) non-existent file and (b) empty file.

**Review by vsajip — CHANGES_REQUESTED** [review]
Thanks for the PR.

**Comment by Agent-Hellboy** [comment]
Hi @vsajip  I have made the requested changes; please review again, 

one more question why doesn't the module has type annotation if this is because you lack people then I would like to do this as the module is very complex and interesting

**Comment by Agent-Hellboy** [comment]
test on Windows is failing 
```
ERROR: test_exception_if_confg_file_is_empty (test.test_logging.ConfigFileTest.test_exception_if_confg_file_is_empty)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\test_logging.py", line 1789, in test_exception_if_confg_file_is_empty
    os.unlink(fn)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\test_empty_eub_b3sw.ini'
```
is windows runner picking random runners in between?
previously I was removing the file in cleanup but now I am removing the file in the test itself but still, the test is failing

**Comment by Agent-Hellboy** [comment]
>test on Windows is failing

Passed

**Comment by Agent-Hellboy** [comment]
Hi @vsajip, please review again

File: Lib/logging/config.py [hunk 3]
```diff
@@ -60,15 +61,24 @@ def fileConfig(fname, defaults=None, disable_existing_loggers=True, encoding=Non
     """
     import configparser
 
+    if isinstance(fname, str):
+        if not os.path.exists(fname):
+            raise FileNotFoundError(f"{fname} doesn't exit")
```
Comment (vsajip) [inline Lib/logging/config.py]:
```suggestion
            raise FileNotFoundError(f"{fname} doesn't exist")
```

File: Lib/logging/config.py [hunk 4]
```diff
@@ -60,15 +61,24 @@ def fileConfig(fname, defaults=None, disable_existing_loggers=True, encoding=Non
...
+        elif not os.path.getsize(fname):
+            raise ValueError(f"{fname} configuration file is empty")
```
Comment (vsajip) [inline Lib/logging/config.py]:
```suggestion
            raise ValueError(f'{fname} is an empty file')
```

File: Lib/logging/config.py [hunk 5]
```diff
@@ -60,15 +61,24 @@ def fileConfig(fname, defaults=None, disable_existing_loggers=True, encoding=Non
...
-            cp.read(fname, encoding=encoding)
+        try:
+            cp = configparser.ConfigParser(defaults)
+            if hasattr(fname, 'readline'):
+                cp.read_file(fname)
+            else:
+                encoding = io.text_encoding(encoding)
+                cp.read(fname, encoding=encoding)
+        except configparser.ParsingError as e:
+            raise ValueError(f"{fname} is invalid, errored with: {e}")
```
Comment (vsajip) [inline Lib/logging/config.py]:
```suggestion
            raise ValueError(f'{fname} is invalid: {e}')
```

File: Lib/test/test_logging.py [hunk 6]
```diff
@@ -1756,6 +1756,42 @@ def test_config_set_handler_names(self):
...
+            """
+
+        file = io.StringIO(textwrap.dedent(test_config))
+        self.assertRaises(ValueError, logging.config.fileConfig, file)
+
+    def test_exception_if_confg_file_is_empty(self):
+        fd, fn = tempfile.mkstemp(prefix='test_empty_', suffix='.ini')
+        os.close(fd)
+        self.assertRaises(ValueError, logging.config.fileConfig, fn)
+        os.unlink(fn)
```
Comment (vsajip) [inline Lib/test/test_logging.py]:
```suggestion
        os.remove(fn)
```

File: Lib/test/test_logging.py [hunk 7]
```diff
@@ -1756,6 +1756,42 @@ def test_config_set_handler_names(self):
...
+
+    def test_exception_if_confg_file_does_not_exists(self):
```
Comment (vsajip) [inline Lib/test/test_logging.py]:
```suggestion
    def test_exception_if_config_file_does_not_exist(self):
```

**Review by vsajip — CHANGES_REQUESTED** [review]

**Comment by Agent-Hellboy** [comment]
Hi @vsajip, should we update in the 3.12 doc that an exception will be thrown if the file is empty or invalid?

**Comment by vsajip** [comment]
> should we update in the 3.12 doc that an exception will be thrown if the file is empty or invalid?

Yes, that would be good - as well as nonexistent file - and indicate the types of exceptions thrown.

**Comment by vsajip** [comment]
> one more question why doesn't the module has type annotation 

No particular reason, and I don't have much time right now to review a large PR so don't worry about doing this.

**Comment by Agent-Hellboy** [comment]
> Yes, that would be good - as well as nonexistent file - and indicate the types of exceptions thrown.
sure 

>No particular reason, and I don't have much time right now to review a large PR so don't worry about doing this.

I see, it has a good unit test so type annotation can be avoided, I hear people saying in conferences that type annotations can be avoided if you have good unit tests.

**Comment by Agent-Hellboy** [comment]
I have made the requested changes; please review again

`test__xxsubinterpreters` and `test_threading` is failing in CI, I guess these are not related to the change

File: Doc/library/logging.config.rst [hunk 8]
```diff
@@ -87,6 +87,10 @@ in :mod:`logging` itself) and defining handlers which are declared either in
    provides a mechanism to present the choices and load the chosen
    configuration).
 
+   It will raise :exc:`FileNotFoundError` if file
```
Comment (vsajip) [inline Doc/library/logging.config.rst]:
```suggestion
   It will raise :exc:`FileNotFoundError` if the file
```

File: Doc/library/logging.config.rst [hunk 9]
```diff
@@ -87,6 +87,10 @@ in :mod:`logging` itself) and defining handlers which are declared either in
...
+   doesn't exist and :exc:`ValueError` if file is invalid or
```
Comment (vsajip) [inline Doc/library/logging.config.rst]:
```suggestion
   doesn't exist and :exc:`ValueError` if the file is invalid or
```

File: Doc/library/logging.config.rst [hunk 10]
```diff
@@ -87,6 +87,10 @@ in :mod:`logging` itself) and defining handlers which are declared either in
...
+   empty
```
Comment (vsajip) [inline Doc/library/logging.config.rst]:
```suggestion
   empty.
```

File: Doc/library/logging.config.rst [hunk 11]
```diff
@@ -126,6 +130,10 @@ in :mod:`logging` itself) and defining handlers which are declared either in
     .. versionadded:: 3.10
        The *encoding* parameter is added.
 
+    .. versionadded:: 3.12
+       The exception will be thrown if the file provided
```
Comment (vsajip) [inline Doc/library/logging.config.rst]:
```suggestion
       An exception will be thrown if the provided file
```

File: Doc/whatsnew/3.12.rst [hunk 12]
```diff
@@ -420,6 +420,13 @@ itertools
   tuples where the last batch may be shorter than the rest.
   (Contributed by Raymond Hettinger in :gh:`98363`.)
 
+logging
```
Comment (vsajip) [inline Doc/whatsnew/3.12.rst]: Please undo this change, there's no need to add this bit.

**Review by vsajip — CHANGES_REQUESTED** [review]

**Review by vsajip — APPROVED** [review]

**Comment by Agent-Hellboy** [comment]
Hi, @miss-islington I cherry-picked it on 3.11

File: Doc/library/logging.config.rst [hunk 13]
```diff
@@ -126,6 +130,10 @@ in :mod:`logging` itself) and defining handlers which are declared either in
     .. versionadded:: 3.10
        The *encoding* parameter is added.
 
+    .. versionadded:: 3.12
```
Comment (gpshead) [inline Doc/library/logging.config.rst:133]: Strictly speaking I'd have said versionchanged rather than versionadded.

File: Lib/logging/config.py [hunk 14]
```diff
@@ -60,15 +61,24 @@ def fileConfig(fname, defaults=None, disable_existing_loggers=True, encoding=Non
     """
     import configparser
 
+    if isinstance(fname, str):
+        if not os.path.exists(fname):
+            raise FileNotFoundError(f"{fname} doesn't exist")
+        elif not os.path.getsize(fname):
+            raise ValueError(f'{fname} is an empty file')
```
Comment (gpshead) [inline Lib/logging/config.py:68]: ValueError is generally only used when the value itself is a problem, not when something external to the process such as a filesystem doesn't look right.  Same below.  If in doubt as to what exception to use here, I'd just go with RuntimeError if you don't have an Error exception class for your package already.
